apps/coordinator/ts/health/health.service.ts

```diff
@@ -0,0 +1,164 @@
+import { EMode } from "@maci-protocol/sdk";
+import { Injectable } from "@nestjs/common";
+import { formatEther } from "ethers";
+
+import { accessSync, constants, statSync } from "fs";
+import path from "path";
+
+import { ESupportedNetworks, getSigner } from "../common";
+import { FileService } from "../file/file.service";
+
+import { ICheckRapidsnark, ICheckWalletFunds, ICheckZkeysDirectory, IHealthCheckResponse } from "./types";
+
+/**
+ * Health service functions to check the status of the coordinator service.
+ */
+@Injectable()
+export class HealthService {
+  /**
+   * Create a new instance of the HealthService.
+   *
+   * @param fileService - file service
+   */
+  constructor(private readonly fileService: FileService) {}
+
+  /**
+   * Check if rapidsnark is available and executable.
+   * @returns the rapidsnark executable path and its executability status
+   */
+  checkRapidsnark(): ICheckRapidsnark {
+    const rapidsnarkExecutablePath = process.env.COORDINATOR_RAPIDSNARK_EXE || "";
+    let rapidsnarkIsExecutable = false;
+
+    try {
+      // Check if it's a file and has execute permissions
+      // eslint-disable-next-line no-bitwise
+      accessSync(rapidsnarkExecutablePath, constants.F_OK | constants.X_OK);
+      const stats = statSync(rapidsnarkExecutablePath);
+
+      rapidsnarkIsExecutable = stats.isFile();
+    } catch (error) {
+      rapidsnarkIsExecutable = false;
+    }
+
+    return {
+      rapidsnarkExecutablePath,
+      rapidsnarkIsExecutable,
+    };
+  }
+
+  /**
+   * Check if zkeys directory exists
+   * @returns the existence of the zkeys directory and the required zkeys
+   */
+  checkZkeysDirectory(): ICheckZkeysDirectory {
+    const zkeysDirectoryPath = path.resolve(process.env.COORDINATOR_ZKEY_PATH!);
+    const requiredZkeyNames = [
+      {
+        name: process.env.COORDINATOR_TALLY_ZKEY_NAME,
+        modes: [EMode.QV, EMode.NON_QV],
+      },
+      {
+        name: process.env.COORDINATOR_MESSAGE_PROCESS_ZKEY_NAME,
+        modes: [EMode.QV, EMode.NON_QV, EMode.FULL],
+      },
+      {
+        name: process.env.COORDINATOR_POLL_JOINING_ZKEY_NAME,
+      },
+      {
+        name: process.env.COORDINATOR_POLL_JOINED_ZKEY_NAME,
+      },
+    ];
+
+    let zkeysDirectoryExists = false;
+    const availableZkeys: ICheckZkeysDirectory["availableZkeys"] = {};
+
+    try {
+      const stats = statSync(zkeysDirectoryPath);
+      zkeysDirectoryExists = stats.isDirectory();
+
+      if (zkeysDirectoryExists) {
+        requiredZkeyNames.forEach(({ name, modes }) => {
+          if (!modes) {
+            const zkeyPath = this.fileService.getZkeyFilePaths(name!);
+            availableZkeys[name!] = zkeyPath;
+            return;
+          }
+
+          modes.forEach((mode) => {
+            const zkeyPath = this.fileService.getZkeyFilePaths(name!, mode);
+            availableZkeys[`${name!}_Mode_${String(mode)}`] = zkeyPath;
+          });
```

Comment:
It's better to use `reduce` to build the immutable `availableZkeys` instead of updating it.